src/main/java/com/redhat/devtools/lsp4ij/features/completion/SnippetTemplateLoader.java

```diff
@@ -32,43 +37,126 @@
 public class SnippetTemplateLoader extends AbstractLspSnippetHandler {
 
     private final Template template;
+    private final boolean useTemplateForInvocationOnlySnippet;
 
-    private final List<String> existingVariables;
+    private final List<SnippetTemplateSegment> templateSegments = new LinkedList<>();

```

Comment:
Instead of building the template on-the-fly, we accumulate the exact same information that we'd need to build the template, and we do it in `endSnippet()` instead. That allows us to have full context of the resulting template segments so we can see whether/how it can be transformed/simplified.